cloud/services/groups/client.go

```diff
@@ -39,8 +39,8 @@ type AzureClient struct {
 var _ Client = &AzureClient{}
 
 // NewClient creates a new VM client from subscription ID.
-func NewClient(auth azure.Authorizer) *AzureClient {
-	c := newGroupsClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
+func NewClient(subscriptionID string, auth azure.Authorizer) *AzureClient {
```

Comment:
What if this was a `ScopedAuthorizer interface` or something? Basically, `Authorizer` + `Subscriber`.

Reply:
I think something like this is probably better, I resisted initially because that smells to me like we should abstract scope a bit more?
I also wonder if we could invert our client instantiation. Create an Azure clientset one time per reconcile. something like:
```go
type ClientSet struct {
  sub string
  auth Authorizer
}

func NewClientSet(sub string, auth Authorizer) *ClientSet { ... }

fung (c *ClientSet) Groups() resources.GroupsClient // or the mocked interface
```

we could plausibly then also group the services, which would simplify mocking for tests.

similar example from hashi:
- client builder: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/clients/client.go
- corresponding client: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/redis/client/client.go

Reply:
`ClusterDescriber.GetClientSet() (*ClientSet, error)` or something?

Reply:
added `SubscriptionAuthorizer` for now

Reply:
I like the `ClientSet.Groups()` idea.

Reply:
> ClusterDescriber.GetClientSet() (*ClientSet, error) or something?

I probably would keep it inverted -- maybe something like
```go
func NewClientSet(sub string, auth Authorizer) *ClientSet { ... }

// similar to existing services
type Service interface {
   Reconcile(ctx context.Context) error
   Delete(ctx context.Context) error
}

type ServiceSet struct {
  *ClientSet // probably interface
  Groups() Service
  Networks() Service
  LoadBalancers() Service
  // etc...
}

func NewServiceSet(d clusterDescriber, c *ClientSet // or interface for mocking) *ServiceSet { ... }

// not necessarily defined on service set, but for illustrative purposes on usage
// our reconciler loops would be calling these repeatedly like they do today for individual services
func (s *ServiceSet) ReconcileGroup(ctx context.Context) error {
  return s.Groups().Reconcile()
}

```

Reply:
would need to consider some differentiation for machine vs cluster, but I like grouping things and doing them once anyway 😄

Reply:
@devigned as I'm working through the next steps to achieve #822 I think ClusterDescriber should exclude authorizer, and we can define an interface over scope which is the union of clusterdescriber and authorizer. I can either do that here or in the next pr and use this as a stepping stone.

Reply:
Why do you say ClusterDescriber should exclude authorizer? As we move toward multitenancy, the cluster with have an identity or the default env identity.

Reply:
we use clusterdescriber to build out infrastructure resources to reconcile. we use the identity only for service instantiation. I want to separate these concerns, I kind of like the approach above for that purpose.

It's also why I'd potentially like to put subscription ID somewhere else, because it's part of auth, not part of reconciliation